ecommerce/serializers.py

```diff
@@ -519,11 +533,23 @@ def validate_items(self, items):
 
             if product_id is None:
                 raise ValidationError("Invalid request")
-            if not models.ProductVersion.objects.filter(
-                product__id=product_id
-            ).exists():
+
+            readable_id = None
+            try:
+                product_id = int(product_id)
+            except ValueError:
+                readable_id = product_id.replace(" ", "+")
+
+            if readable_id:
+                product = models.Product.all_objects.filter(
+                    dj_models.Q(courseruns__courseware_id=readable_id)
+                    | dj_models.Q(programs__readable_id=readable_id)
+                ).first()
+            elif product_id:
+                product = models.Product.all_objects.filter(id=product_id).first()
+
+            if not product:
```

Comment:
If you implement the API function above, this block would look something like this:

```python
from django.core.exceptions import ObjectDoesNotExist, MultipleObjectsReturned

try:
  product = get_product_from_querystring_id(product_id)
except (ObjectDoesNotExist, MultipleObjectsReturned) as exc:
  if isinstance(exc, MultipleObjectsReturned):
    log.error("Multiple Products found with identical ids: %s", product_id)
  raise ValidationError(f"Invalid product id {product_id}") from exc
```

This also avoids arbitrarily using `first()` for the Product query using the text id, which I don't think we should be doing. Unless I'm mistaken about something, I don't think we should just pick the first one and assume it's correct if multiple Products have the same text id. We should log an error and deny the request

Reply:
As per domain knowledge, one course-run / program can have only Product so that's why other cases not covered.

Reply:
**As per the model, we can't have multiple products for a course run or program. There will be only one product for each (either course run or progam).
`unique_together = ("content_type", "object_id")`**

I have verified it too on xpro. Here is the screenshot for the reference.

<img width="1424" alt="Screenshot 2020-01-31 at 17 28 50" src="https://user-images.githubusercontent.com/4043989/73539536-65d84980-444f-11ea-9f34-eb365e581853.png">